File: template-assets/javascript/test_converter.js
const assert = require('assert');
const dsUtils = require('./ds_utils');
const { TreeNode, Graph, ListNode } = dsUtils;
const { listyToType } = require('./converter');

describe('listyToType Integration Tests', function() {
    it('should convert to TreeNode', function() {
        const stringyInput = "[1, 2, 3, 1, 1, 4, 5]";
        const abstractType = {
            type: "TreeNode",
            typeChildren: {
                type: "Integer"
            }
        };

        const tree = listyToType(stringyInput, abstractType);
        assert(tree instanceof TreeNode);
        assert.deepStrictEqual(dsUtils.exportTree(tree), [1, 2, 3, 1, 1, 4, 5]);
    });

    it('should convert to Graph', function() {
        const stringyInput = "[[1, 2], [2, 3], [3, 1]]";
        const abstractType = {
            type: "Graph",
            typeChildren: {
                type: "Integer"
            }
        };

        const graph = listyToType(stringyInput, abstractType);
        console.log(graph);
        console.log(dsUtils.exportGraph(graph));
        assert(graph instanceof Graph);
        assert.deepStrictEqual(dsUtils.exportGraph(graph), [[1, 2], [2, 3], [3, 1]]);
    });

    it('should convert to ListNode', function() {
        const stringyInput = "[1, 2, 3, 4]";
        const abstractType = {
            type: "ListNode",
            typeChildren: {
                type: "Integer"
            }
        };

        const linkedList = listyToType(stringyInput, abstractType);
        assert(linkedList instanceof ListNode);
        assert.deepStrictEqual(dsUtils.exportLinkedList(linkedList), [1, 2, 3, 4]);
    });

    it('should convert to Array', function() {
        const stringyInput = "[1, 2, 3, 4]";
        const abstractType = {
            type: "Array",
            typeChildren: {
                type: "Integer"
            }
        };

        const array = listyToType(stringyInput, abstractType);
        assert(Array.isArray(array));
        assert.deepStrictEqual(array, [1, 2, 3, 4]);
    });

    it('should convert to Matrix', function() {
        const stringyInput = "[[1, 2], [3, 4], [5, 6]]";
        const abstractType = {
            type: "Matrix",
            typeChildren: {
                type: "Integer"
            }
        };

        const matrix = listyToType(stringyInput, abstractType);
        assert(Array.isArray(matrix));
        assert.deepStrictEqual(matrix, [[1, 2], [3, 4], [5, 6]]);
    });

    it('should throw error for invalid type', function() {
        const stringyInput = "[1, 2, 3]";
        const abstractType = { type: "Unknown" };
        assert.throws(() => listyToType(stringyInput, abstractType), Error);
    });
});

File: template-assets/python/test_converter.py
import unittest
import ds_utils
from ds_utils import TreeNode, Graph, ListNode
from converter import listy_to_type  # Import your `listy_to_type` function here


class TestListyToTypeIntegration(unittest.TestCase):
    def test_tree_node(self):
        stringy_input = "[1, 2, 3, 1, 1, 4, 5]"
        abstract_type = {
            "type": "TreeNode",
            "type_children": {
                "type": "Integer"
            }
        }

        # Test parsing and exporting TreeNode
        tree = listy_to_type(stringy_input, abstract_type)
        self.assertTrue(isinstance(tree, TreeNode))
        self.assertEqual(ds_utils.export_tree(tree), [1, 2, 3, 1, 1, 4, 5])

    def test_graph(self):
        stringy_input = "[[1, 2], [2, 3], [3, 1]]"
        abstract_type = {
            "type": "Graph",
            "type_children": {
                "type": "Integer"
            }
        }

        # Test parsing and exporting Graph
        graph = listy_to_type(stringy_input, abstract_type)
        self.assertTrue(isinstance(graph, Graph))
        self.assertEqual(ds_utils.export_graph(graph), [[1, 2], [2, 3], [3, 1]])

    def test_list_node(self):
        stringy_input = "[1, 2, 3, 4]"
        abstract_type = {
            "type": "ListNode",
            "type_children": {
                "type": "Integer"
            }
        }

        # Test parsing and exporting ListNode
        linked_list = listy_to_type(stringy_input, abstract_type)
        self.assertTrue(isinstance(linked_list, ListNode))
        self.assertEqual(ds_utils.export_linked_list(linked_list), [1, 2, 3, 4])

    def test_array(self):
        stringy_input = "[1, 2, 3, 4]"
        abstract_type = {
            "type": "Array",
            "type_children": {
                "type": "Integer"
            }
        }

        # Test parsing and exporting Array
        array = listy_to_type(stringy_input, abstract_type)
        self.assertTrue(isinstance(array, list))
        self.assertEqual(array, [1, 2, 3, 4])

    def test_matrix(self):
        stringy_input = "[[1, 2], [3, 4], [5, 6]]"
        abstract_type = {
            "type": "Matrix",
            "type_children": {
                "type": "Integer"
            }
        }

        # Test parsing and exporting Matrix
        matrix = listy_to_type(stringy_input, abstract_type)
        self.assertTrue(isinstance(matrix, list))
        self.assertEqual(matrix, [[1, 2], [3, 4], [5, 6]])

    # def test_nested_tree_in_graph(self):
    #   # Input: Graph of trees
    #     stringy_input = "[[[1, 2, 1], [2, 1, 1]], [[2, 1, 1], [3, 1, 1]]]"
    #     abstract_type = {
    #         "type": "Graph",
    #         "type_children": {
    #             "type": "TreeNode",
    #             "type_children": {
    #                 "type": "Integer"
    #             }
    #         }
    #     }


    #     # Test parsing and exporting nested TreeNode in Graph
    #     graph = listy_to_type(stringy_input, abstract_type)
    #     self.assertTrue(isinstance(graph, Graph))
    #     exported_graph = ds_utils.export_graph(graph)
    #     self.assertEqual(
    #         exported_graph,
    #         [[TreeNode(1), TreeNode(2)]]
    #     )

    def test_invalid_type(self):
        stringy_input = "[1, 2, 3]"
        abstract_type = {"type": "Unknown"}
        with self.assertRaises(ValueError):
            listy_to_type(stringy_input, abstract_type)


if __name__ == "__main__":
    unittest.main()
